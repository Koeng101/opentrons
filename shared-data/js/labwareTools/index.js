// @flow
import assignId from './assignId'
import roundTo from 'round-to'

type Metadata = {
  name: string,
  displayCategory: string,
  displayVolumeUnits?: string,
  displayLengthUnits?: string,
}

type Dimensions = {
  overallLength: number,
  overallWidth: number,
  overallHeight: number,
  offsetX: number,
  offsetY: number,
  offsetZ: number,
}

type Vendor = {
  sku?: string,
  vendor?: string,
}

// 1. Valid pipette type for a container (i.e. is there multi channel access?)
// 2. Is the container a tiprack?
type Params = {
  format: string,
  isTiprack: boolean,
  tipLength?: number,
  wellShape?: string,
}

type Well = {
  depth: number,
  diameter?: number,
  length?: number,
  width?: number,
  totalLiquidVolume?: number,
}

type Schema = {
  otId: number,
  deprecated: boolean,
  metadata: Metadata,
  dimensions: Dimensions,
  parameters: Params,
  vendor?: Vendor,
  ordering: Array<Array<string>>,
  wells: {[wellName: string]: Well},
}

function determineOrdering (grid: [number, number]): Array<Array<string>> {
  var ordering = []
  var rows = grid[0]
  var cols = grid[1]
  var A = 'A'

  var r
  var c

  for (c = 0; c < cols; c++) {
    let colOrdering = []
    for (r = 0; r < rows; r++) {
      var char = String.fromCharCode(A.charCodeAt(0) + r)
      var wellName = char + (1 + c).toString()
      colOrdering.push(wellName)
    }
    ordering.push(colOrdering)
  }

  return ordering
}
// Private helper function to return individual well output
function calculateCoordinates (
  well: Well,
  ordering: Array<Array<string>>,
  spacing: [number, number]
): {[wellName: string]: Well} {
  let wells = {}
  var rowSpacing = spacing[0]
  var colSpacing = spacing[1]
  var col
  var row

  for (col = 0; col < ordering.length; col++) {
    var rowLength = ordering[col].length
    for (row = 0; row < rowLength; row++) {
      var x = roundTo(col * colSpacing, 2)
      var y = roundTo((rowLength - row - 1) * rowSpacing, 2)
      var z = 0
      wells[ordering[col][row]] = {...well, x, y, z}
    }
  }

  return wells
}

export type RegularLabwareProps = {|
  metadata: Metadata,
  parameters: Params,
  dimensions: Dimensions,
  grid: [number, number],
  spacing: [number, number],
  well: Well,
  vendor?: Vendor,
|}

// Make function public via export function
// Need slot offset
// well spacing
// use display category to do logic checks (tiprack has tip length etc)
// rows/cols
// customization can contain nothing OR ??
export function createRegularLabware (props: RegularLabwareProps): Schema {
  const ordering = determineOrdering(props.grid)

  const definition: Schema = {
    ordering,
    otId: assignId(),
    deprecated: false,
    metadata: props.metadata,
    dimensions: props.dimensions,
    parameters: props.parameters,
    wells: calculateCoordinates(props.well, ordering, props.spacing),
  }

  if (props.vendor) definition.vendor = props.vendor

  return definition
}
