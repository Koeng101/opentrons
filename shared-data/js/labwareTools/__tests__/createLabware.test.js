import Ajv from 'ajv'
import exampleLabware1 from './example.json'
import exampleLabware2 from './example2.json'
import labwareSchema from '../../../labware-json-schema/labware-schema.json'
import omit from 'lodash/omit'
import range from 'lodash/range'
import {createRegularLabware} from '../index.js'

jest.mock('../assignId', () => jest.fn(() => 'mock-id'))

const ajv = new Ajv({
  allErrors: true,
  jsonPointers: true,
})

const validate = ajv.compile(labwareSchema)

describe('createLabware', () => {
  let labware1
  let labware2
  let well1
  let well2

  beforeEach(() => {
    well1 = omit(exampleLabware1.wells.A1, ['x', 'y', 'z'])
    well2 = omit(exampleLabware2.wells.A1, ['x', 'y', 'z'])

    labware1 = createRegularLabware({
      metadata: exampleLabware1.metadata,
      parameters: exampleLabware1.parameters,
      dimensions: exampleLabware1.dimensions,
      grid: [1, 2],
      spacing: [10, 10],
      well: well1,
      vendor: exampleLabware1.vendor,
    })

    labware2 = createRegularLabware({
      metadata: exampleLabware2.metadata,
      parameters: exampleLabware2.parameters,
      dimensions: exampleLabware2.dimensions,
      grid: [3, 2],
      spacing: [9, 9],
      well: well2,
    })
  })

  afterEach(() => {
    jest.clearAllMocks()
  })

  test('snapshot tests', () => {
    expect(labware1).toMatchSnapshot()
    expect(labware2).toMatchSnapshot()
  })

  test('generated labware passes schema', () => {
    const valid = validate(exampleLabware1)
    const validationErrors = validate.errors

    expect(validationErrors).toBe(null)
    expect(valid).toBe(true)
  })

  test('id is from assignId', () => {
    expect(labware1.otId).toBe('mock-id')
    expect(labware2.otId).toBe('mock-id')
  })

  test('ordering generates as expected', () => {
    expect(exampleLabware2.ordering).toEqual(labware2.ordering)
  })

  test('well XYZ generates correctly', () => {
    const spacing = [11.8, 12.1]
    const grid = [8, 12]

    const labware3 = createRegularLabware({
      metadata: exampleLabware2.metadata,
      paramenters: exampleLabware2.parameters,
      dimensions: exampleLabware2.dimensions,
      grid,
      spacing,
      well: well2,
    })

    const expectedXByCol = range(0, grid[1] * spacing[1], spacing[1])
    const expectedYByRow = range(0, grid[0] * spacing[0], spacing[0]).reverse()

    labware3.ordering.forEach((column, cIndex) => {
      column.forEach((wellName, rIndex) => {
        const well = labware3.wells[wellName]
        expect(well.x).toBeCloseTo(expectedXByCol[cIndex], 2)
        expect(well.y).toBeCloseTo(expectedYByRow[rIndex], 2)
        expect(well.z).toBeCloseTo(0, 2)
      })
    })
  })
})
