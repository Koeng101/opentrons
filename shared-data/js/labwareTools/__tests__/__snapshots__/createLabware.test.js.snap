// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createLabware snapshot tests 1`] = `
Object {
  "deprecated": false,
  "dimensions": Object {
    "offsetX": 10,
    "offsetY": 10,
    "offsetZ": 5,
    "overallHeight": 50,
    "overallLength": 50,
    "overallWidth": 50,
  },
  "metadata": Object {
    "displayCategory": "wellPlate",
    "displayLengthUnits": "mm",
    "displayVolumeUnits": "uL",
    "name": "fake labware",
  },
  "ordering": Array [
    Array [
      "A1",
    ],
    Array [
      "A2",
    ],
  ],
  "otId": "mock-id",
  "parameters": Object {
    "format": "96Standard",
    "isTiprack": false,
    "wellShape": "circular",
  },
  "vendor": Object {
    "sku": "t40u9sernisofsea",
    "vendor": "opentrons",
  },
  "wells": Object {
    "A1": Object {
      "depth": 40,
      "diameter": 30,
      "totalLiquidVolume": 100,
      "x": 0,
      "y": 0,
      "z": 0,
    },
    "A2": Object {
      "depth": 40,
      "diameter": 30,
      "totalLiquidVolume": 100,
      "x": 10,
      "y": 0,
      "z": 0,
    },
  },
}
`;

exports[`createLabware snapshot tests 2`] = `
Object {
  "deprecated": false,
  "dimensions": Object {
    "offsetX": 10,
    "offsetY": 10,
    "offsetZ": 5,
    "overallHeight": 50,
    "overallLength": 50,
    "overallWidth": 50,
  },
  "metadata": Object {
    "displayCategory": "wellPlate",
    "displayLengthUnits": "mm",
    "displayVolumeUnits": "uL",
    "name": "fake labware 2",
  },
  "ordering": Array [
    Array [
      "A1",
      "B1",
      "C1",
    ],
    Array [
      "A2",
      "B2",
      "C2",
    ],
  ],
  "otId": "mock-id",
  "parameters": Object {
    "format": "irregular",
    "isTiprack": false,
    "wellShape": "circular",
  },
  "wells": Object {
    "A1": Object {
      "depth": 40,
      "diameter": 30,
      "totalLiquidVolume": 100,
      "x": 0,
      "y": 18,
      "z": 0,
    },
    "A2": Object {
      "depth": 40,
      "diameter": 30,
      "totalLiquidVolume": 100,
      "x": 9,
      "y": 18,
      "z": 0,
    },
    "B1": Object {
      "depth": 40,
      "diameter": 30,
      "totalLiquidVolume": 100,
      "x": 0,
      "y": 9,
      "z": 0,
    },
    "B2": Object {
      "depth": 40,
      "diameter": 30,
      "totalLiquidVolume": 100,
      "x": 9,
      "y": 9,
      "z": 0,
    },
    "C1": Object {
      "depth": 40,
      "diameter": 30,
      "totalLiquidVolume": 100,
      "x": 0,
      "y": 0,
      "z": 0,
    },
    "C2": Object {
      "depth": 40,
      "diameter": 30,
      "totalLiquidVolume": 100,
      "x": 9,
      "y": 0,
      "z": 0,
    },
  },
}
`;
